### PR TITLE
[RFR] Upgrade React-router

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

File: packages/ra-core/src/auth/useLogout.ts [hunk 1]
```diff
@@ -40,42 +40,38 @@ const useLogout = (): Logout => {
 
     const store = useStore();
     const dispatch = useDispatch();
+    const history = useHistory();
 
     const logout = useCallback(
         (params = {}, redirectTo = defaultAuthParams.loginUrl) =>
             authProvider.logout(params).then(redirectToFromProvider => {
                 dispatch(clearState());
                 const currentLocation = store.getState().router.location;
```
Comment (fzaninotto) [inline packages/ra-core/src/auth/useLogout.ts]: Can't you get rid of this one redux usage?

File: packages/ra-core/src/auth/useLogout.ts [hunk 2]
```diff
@@ -40,42 +40,38 @@ const useLogout = (): Logout => {
...
+                });
                 return redirectToFromProvider;
             }),
-        [authProvider, store, dispatch]
+        [authProvider, store, history, dispatch]
     );
 
     const logoutWithoutProvider = useCallback(
         _ => {
             const currentLocation = store.getState().router.location;
```
Comment (fzaninotto) [inline packages/ra-core/src/auth/useLogout.ts]: and here, too.

File: packages/ra-ui-materialui/src/field/ArrayField.spec.js [hunk 3]
```diff
@@ -60,9 +61,11 @@ describe('<ArrayField />', () => {
             </ArrayField>
         );
         const wrapper = mount(
-            <Provider store={createStore(() => ({}))}>
-                <Dummy />
-            </Provider>
+            <MemoryRouter>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ArrayField.spec.js]: you can pretty much use the TestContext here

**Comment by fzaninotto** [comment]
also, do we still use `connected-react-router` for anything else than sagas?

**Comment by fzaninotto** [comment]
And there is the FIXME in the CloneButton, can we get rid of it?

File: packages/ra-ui-materialui/src/field/ArrayField.spec.js (hunk 3 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/field/ArrayField.spec.js]: I cleanup those tests and rewrote them using rtl. One of them was useless

**Comment by fzaninotto** [comment]
> Only place left is in the saving reducer.

I think we can get rid of this one, as the `useUpdate` and `useCreate` hook pass their own `loading` state. We'd need to pass it down from  Edit to SimpleForm instead of selecting the value in SimpleForm.

As it's a BC break, we'd better do it now. 

> However, I'm not sure we should pass the record in the state. Having it in the URL is a good thing imo: explicit, allow to refresh the page, copy/paste the url, etc.

I agree. Can you check that using the `state` is never documented anywhere and remove the FIXME then?

**Pull request merged**